controllers/secret_controller.go

```diff
@@ -0,0 +1,370 @@
+		if err != nil {
+			logger.Error(err, "Failed to fetch remote client for managed cluster", "ReplicatedSecret", req.NamespacedName, "TargetContext", target)
+			return ctrl.Result{Requeue: true}, err
+		}
+
+		cond := api.ReplicationCondition{
+			Cluster: target.K8sContextName,
+			Type:    api.ReplicationDone,
+		}
+
+	TargetSecrets:
+		// Iterate all the matching secrets
+		for i := range secrets {
+			sec := &secrets[i]
+			fetchedSecret := &corev1.Secret{}
+			if err = remoteClient.Get(ctx, types.NamespacedName{Name: sec.Name, Namespace: sec.Namespace}, fetchedSecret); err != nil {
+				if errors.IsNotFound(err) {
+					logger.Info("Copying secret to target cluster", "Secret", sec.Name, "TargetContext", target)
+					// Create it
+					copiedSecret := sec.DeepCopy()
+					copiedSecret.ResourceVersion = ""
+					copiedSecret.OwnerReferences = []metav1.OwnerReference{}
+					if err = remoteClient.Create(ctx, copiedSecret); err != nil {
+						logger.Error(err, "Failed to sync secret to target cluster", "Secret", copiedSecret.Name, "TargetContext", target)
+						break TargetSecrets
+					}
+					continue
+				}
+				logger.Error(err, "Failed to fetch secret from target cluster", "Secret", fetchedSecret.Name, "TargetContext", target)
+				break TargetSecrets
+			}
+
+			if fetchedSecret.Immutable != nil && *fetchedSecret.Immutable {
+				err := fmt.Errorf("target secret is immutable")
+				logger.Error(err, "Failed to modify target secret, secret is set to immutable", "Secret", fetchedSecret.Name, "TargetContext", target)
+				break TargetSecrets
+			}
+
+			if requiresUpdate(sec, fetchedSecret) {
+				logger.Info("Modifying secret in target cluster", "Secret", sec.Name, "TargetContext", target)
+				syncSecrets(sec, fetchedSecret)
+				if err = remoteClient.Update(ctx, fetchedSecret); err != nil {
+					logger.Error(err, "Failed to sync target secret for matching payloads", "Secret", fetchedSecret.Name, "TargetContext", target)
+					break TargetSecrets
+				}
+			}
+		}
+		if err != nil {
+			cond.Status = corev1.ConditionFalse
+		} else {
+			cond.Status = corev1.ConditionTrue
+		}
+
+		timeNow := metav1.Now()
+		cond.LastTransitionTime = &timeNow
+		rsec.Status.Conditions = append(rsec.Status.Conditions, cond)
+	}
+
+	// Update the ReplicatedSecret's Status
+	err = localClient.Status().Patch(ctx, rsec, patch)
+	if err != nil {
+		logger.Error(err, "Failed to update replicated secret last transition time", "ReplicatedSecret", req.NamespacedName)
+		return ctrl.Result{Requeue: true}, err
+	}
+
+	// If any cluster had failed state, retry
+	for _, cond := range rsec.Status.Conditions {
+		if cond.Status == corev1.ConditionFalse {
+			return ctrl.Result{Requeue: true}, fmt.Errorf("replication failed")
+		}
+	}
+	return ctrl.Result{}, err
+}
+
+func requiresUpdate(source, dest client.Object) bool {
+	if source.GetGeneration() != 0 {
+		// Compare using generation
+		return source.GetGeneration() > dest.GetGeneration()
```

Comment:
Yeah, it should be safe (since we're the ones making the generation updates), although just like with resource hash annotations, the user can mess things up. 

Though, generations did not work on my envtest (they were always 0).

Reply:
>Yeah, it should be safe (since we're the ones making the generation updates)

We can't assume though that we are the only ones making the updates.